File: python/64.Minimum Path Sum.py
class Solution:
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m=len(grid)
        if m==0:
        	return 0
        n=len(grid[0])
        dp=[[0]*n for i in range(m)]
        for i in range(m):
        	for j in range(n):
        		if i==0 and j==0:
        			dp[i][j]=grid[0][0]
        		elif i==0:
        			dp[i][j]=grid[i][j]+dp[i][j-1]
        		elif j==0:
        			dp[i][j]=grid[i][j]+dp[i-1][j]
        		else:
        			dp[i][j]=grid[i][j]+min(dp[i-1][j],dp[i][j-1])
        return dp[m-1][n-1]


print(Solution().minPathSum([[1,3,1],[1,5,1]]))

File: cpp/64.Minimum Path Sum.cpp
# include <vector>
using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<long>>dp(m);
        for (int i=0; i<m; i++)
            dp[i].resize(n);
        for (int i=0; i<m; i++){
            if (i==0)
                dp[i][0] = grid[i][0];
            else
                dp[i][0] = dp[i-1][0] + grid[i][0];
            for (int j=0; j<n; j++){
                if (j == 0)
                    dp[0][j] = grid[0][j];
                else
                    dp[0][j] = dp[0][j-1] + grid[0][j];
                if (i != 0 and j!=0)
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j];
            }
        }
        return dp[m-1][n-1];
    }
};